### PR TITLE
added gruppo 3 data-link structure

### PR DESCRIPTION
**Pull request by MattiaFanan-uni**
this is gruppo3 data-link layer and interfaces

### CODE REVIEW

**Comment by CremaLuca** [comment]
21 files to review all at once 🥵

File: smslibrary/src/main/java/com/eis/smslibrary/manager/SMSManager.java [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+    {
+        return new ArrayList<SMSDataUnit>();
+    }
+
+
+    /**
+     * future
+     * save a pendingmessge in database
+     */
+    //TODO
```
Comment (CremaLuca) [inline smslibrary/src/main/java/com/eis/smslibrary/manager/SMSManager.java]: What? Why bothering having us review your code if you still have TO-DOs? Finish it, then we can think about merging

File: smslibrary/src/main/java/com/eis/communication/Header.java [hunk 2]
```diff
@@ -0,0 +1,90 @@
+package com.eis.communication;
+
+/**
+ * Header abstract class
```
Comment (CremaLuca) [inline smslibrary/src/main/java/com/eis/communication/Header.java]: Be more explicative when adding specifications on top of a class: what does this class represent? How should it be used?

File: smslibrary/src/main/java/com/eis/communication/Header.java [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+    }
+
+    /**
+     * Set destination peer in the Header
+     * @param destinationPeer to set in the Header
+     * @return true if destinationPeer is valid, false if destinationPeer is not valid
+     */
+    public boolean setDestinationPeer(P destinationPeer){
+        if (destinationPeer == null){
+            //peers can't be both null
```
Comment (CremaLuca) [inline smslibrary/src/main/java/com/eis/communication/Header.java]: What's the point of having two peers then? Can't you just have one and treat it as the sender/receiver?

File: smslibrary/src/main/java/com/eis/communication/Payload.java [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+ * @author Mattia Fanan
+ *
+ * @param T Payload data data-type
+ */
+public abstract class Payload<T> {
+    protected T data;
+
+    /**
+     * Create a payload
+     * @param data Data of the Payload
```
Comment (CremaLuca) [inline smslibrary/src/main/java/com/eis/communication/Payload.java]: Don't repeat yourself on param specifications
do something like `@param data of the payload`

File: smslibrary/src/main/java/com/eis/communication/Payload.java [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+
+        this.data = data;
+        return true;
+    }
+
+    /**
+     * Check if Payload is valid
+     * @return true if Payload is valid, false if Payload is not valid
+     */
+    public boolean isValid(){
```
Comment (CremaLuca) [inline smslibrary/src/main/java/com/eis/communication/Payload.java]: Can I have an invalid instance of Payload? If no, what is the purpose of this method? Will it ever return false?

File: smslibrary/src/main/java/com/eis/communication/Peer.java [hunk 6]
```diff
@@ -1,25 +1,58 @@
 package com.eis.communication;
 
-import androidx.annotation.NonNull;
-
-import java.io.Serializable;
+import com.eis.communication.exception.InvalidPeerException;
 
 /**
- * Represents a user in the network
```
Comment (CremaLuca) [inline smslibrary/src/main/java/com/eis/communication/Peer.java]: Why removing this?

File: smslibrary/src/main/java/com/eis/communication/Peer.java [hunk 7]
```diff
@@ -1,25 +1,58 @@
...
- * @param <A> The type of address used by the network
+ * @author Mattia Fanan
+ * abstarcton of peer class
+ * @param T type of address
  */
-public interface Peer<A> extends Serializable {
```
Comment (CremaLuca) [inline smslibrary/src/main/java/com/eis/communication/Peer.java]: Why removing Serializable?

File: smslibrary/src/main/java/com/eis/communication/Peer.java [hunk 8]
```diff
@@ -1,25 +1,58 @@
...
+public abstract class Peer<T> {
```
Comment (CremaLuca) [inline smslibrary/src/main/java/com/eis/communication/Peer.java]: Can't we have a full set of interfaces and then have something like `AbstractPeer<T>`? It's clearer.

File: smslibrary/src/main/java/com/eis/communication/Payload.java (hunk 5 above)
Comment (MattiaFanan-uni) [reply smslibrary/src/main/java/com/eis/communication/Payload.java]: 👍
I have removed all isValid methods
(In my repo)

File: smslibrary/src/main/java/com/eis/communication/Header.java (hunk 3 above)
Comment (MattiaFanan-uni) [reply smslibrary/src/main/java/com/eis/communication/Header.java]: 👍
Maybe in future I can delete one peer but now I’m not much sure is the right thing to do

**Comment by MattiaFanan-uni** [comment]
if Someone wants can block this pull request👍

File: smslibrary/src/main/java/com/eis/communication/Peer.java (hunk 8 above)
Comment (MattiaFanan-uni) [reply smslibrary/src/main/java/com/eis/communication/Peer.java]: I’m not fully convinced
Everyone i found in the net says that interfaces and abstract classes are more or less the same
You use abstract for implementing some default behavior 
And interfaces if you don’t know the default behavior a priori

**Comment by CremaLuca** [comment]
Closing because of a misunderstanding from the author

**Pull request closed without merging**